hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala

```diff
@@ -39,17 +39,17 @@ object GenotypeFunctions extends RegistryFunctions {
         m2 - m)
     }
 
-    registerCode[Long]("dosage", TArray(tv("N", "float64")), TFloat64,  (pt: PType) => PFloat64()) { case (r, rt, (gpPType, gpOff)) =>
-      val gpPArray = coerce[PArray](gpPType)
+    registerPCode("dosage", TArray(tv("N", "float64")), TFloat64,  (pt: PType) => PFloat64()
+    ) { case (r, rt, gp: PIndexableCode) =>
+      val code = EmitCodeBuilder.scopedCode[Double](r.mb) { cb =>
+        val gpv: PIndexableValue = gp.memoize(cb, "dosage_gp")
+        cb.ifx(gpv.loadLength() > 3,
+          Code._fatal[Unit](const("length of gp array must be 3, got ").concat(gpv.loadLength().toS)))
 
-      Code.memoize(gpOff, "dosage_gp") { gp =>
-        Code.memoize(gpPArray.loadLength(gp), "dosage_len") { len =>
-        len.cne(3).mux(
-          Code._fatal[Double](const("length of gp array must be 3, got ").concat(len.toS)),
-          Region.loadDouble(gpPArray.elementOffset(gp, 3, 1)) +
-            Region.loadDouble(gpPArray.elementOffset(gp, 3, 2)) * 2.0)
-        }
+        gpv.loadElement(cb, 1).pc.tcode[Double] +
+          gpv.loadElement(cb, 2).pc.tcode[Double] * const(2.0)
```

Comment:
@cseed All of the test failures are related to this change. There must be some contract I am violating, possibly due to just using IEmitCode's PCode. Help?
```
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at is.hail.lir.Classx$$anonfun$asBytes$3$$anonfun$apply$3.apply(X.scala:84)
	at is.hail.lir.Classx$$anonfun$asBytes$3$$anonfun$apply$3.apply(X.scala:83)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at is.hail.lir.Classx$$anonfun$asBytes$3.apply(X.scala:83)
	at is.hail.lir.Classx$$anonfun$asBytes$3.apply(X.scala:81)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at is.hail.lir.Classx.asBytes(X.scala:81)
	at is.hail.asm4s.ClassBuilder.classAsBytes(FunctionBuilder.scala:233)
	at is.hail.asm4s.ModuleBuilder$$anonfun$classesBytes$1.apply(FunctionBuilder.scala:84)
	at is.hail.asm4s.ModuleBuilder$$anonfun$classesBytes$1.apply(FunctionBuilder.scala:84)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
	at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:59)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:104)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:48)
	at scala.collection.TraversableOnce$class.to(TraversableOnce.scala:310)
	at scala.collection.AbstractIterator.to(Iterator.scala:1334)
	at scala.collection.TraversableOnce$class.toBuffer(TraversableOnce.scala:302)
	at scala.collection.AbstractIterator.toBuffer(Iterator.scala:1334)
	at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:289)
	at scala.collection.AbstractIterator.toArray(Iterator.scala:1334)
	at is.hail.asm4s.ModuleBuilder.classesBytes(FunctionBuilder.scala:85)
	at is.hail.expr.ir.EmitClassBuilder.resultWithIndex(EmitFunctionBuilder.scala:537)
	at is.hail.expr.ir.WrappedEmitClassBuilder$class.resultWithIndex(EmitFunctionBuilder.scala:98)
	at is.hail.expr.ir.EmitFunctionBuilder.resultWithIndex(EmitFunctionBuilder.scala:903)
```